### PR TITLE
[Inductor] inplace padding

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145325
* __->__ #140249

https://github.com/pytorch/pytorch/issues/139865

This PR may change the semantic of constant_pad_nd from 'clone' to 'view'. I tried a few tests to do inplace update. Looks like thanks to functionalization, this works fine.

Perf for `test_linear_and_cel`:
```
# TORCHINDUCTOR_INPLACE_PADDING=0 DO_PERF_TEST=1 python test/inductor/test_inplace_padding.py -k test_linear_and_cel
inductor_config.inplace_padding=False ms=83.311

# TORCHINDUCTOR_INPLACE_PADDING=1 DO_PERF_TEST=1 python test/inductor/test_inplace_padding.py -k test_linear_and_cel
inductor_config.inplace_padding=True ms=79.827
```

The saving is about 4ms (slightly less since we need fill 0 for the padding area). Similar savings for llm.c.
- Without the feature: 182.151ms per batch, 180.9K tokens/s
- With the feature:  178.278ms per batch, 183.9K tokens/s. There are 3K tokens/s increase.

Perf test shows compilation time regression. . I'm not sure if that's real. Will debug more. But a good thing is, there is no accuracy failure: [link](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2004%20Nov%202024%2020%3A23%3A22%20GMT&stopTime=Mon%2C%2011%20Nov%202024%2020%3A23%3A22%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/shunting314/186/head&lCommit=03fd924ff382958daf5055dc8425d279e4e10a1e&rBranch=main&rCommit=c03324de2dfbbf0006818c86b88c92a3378f46b7) .

UPDATE: Perf test regression seems to be not real. Here is a rerun [link](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2007%20Nov%202024%2001%3A29%3A55%20GMT&stopTime=Thu%2C%2021%20Nov%202024%2001%3A29%3A55%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/shunting314/186/head&lCommit=7e2c8e5d9256ac06205e7cd5e740c9e20ce804d0&rBranch=main&rCommit=565a7942eee1ddc23067cdbae597443d0f2290a0). Our dashboard is not that reliable recently due to AWS migration.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D68340248](https://our.internmc.facebook.com/intern/diff/D68340248)

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Review by Chillee — CHANGES_REQUESTED** [review]
Why does this optimization have to be restricted to mm? Isn't this just always a good idea?

**Comment by shunting314** [comment]
> Why does this optimization have to be restricted to mm? Isn't this just always a good idea?

It will be less promising if the padding op can be fused to downstream kernels.

Eg., Let's say we have tensor `A` to be padded. Call the padding tensor `b` (it's a slice). Call the padded tensor A+b . (b is much smaller than A). And padded tensor is needed by the downstream kernel K.

- if the padding kernel can be fused with kernel K: This PR will introduce extra memory access to clear tensor b and read it. The write of tensor b can be a bit slow due to the memory layout
- if the padding kernel can NOT be fused with kernel K: then it would always be good to do inplace padding. Without the PR, the membw needed for padding is (A.nbytes + A.nbytes + b.nbytes). With the PR, the membw needed to access the padded tensor is (b.nbytes + A.nbytes + b.nbytes). So the saving is (A.nbytes - b.nbytes). The saving can be slightly worse since clearing tensor b can be slow due to its layout.

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -3814,12 +3815,91 @@ def loader(idx):
...
+
+    if not ir.is_storage_and_layout(x):
+        return None
+
+    x.freeze_layout()
+
+    _, layout = ir.as_storage_and_layout(x)
+    strides = layout.stride
+    if strides[1] != 1:
+        return None
```
Comment (eellison) [inline torch/_inductor/lowering.py:3862]: I'm a bit confused about this. 
if we have a tensor of size `[62, 2048]` and we are padding `[0, 2, 0, 0]` the input tensor will be allocated for size [62, 2048]. 

We can't just zero_ `[63:64, :]` because we haven't actually made that allocation. We don't own this memory. 

This is also basically kind of a rewrite of the aten.cat optimization. Which I think we should be doing for mm planning but @oniononion36 noticed that using constant_pad_nd speeds up the padding compared to cat. 		We should just fix the slowdown, and there is a repro D64571598, although i'm not sure if i'm using it correctly.

**Review by eellison — COMMENTED** [review]
Not sure this is correct.

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/lowering.py:3862]: The semantic of the padding parameter is kind of 'reversed'. The semantic of padding [0, 2, 0, 0] to a [62, 2048] tensor is to add 2 elements after each row (not adding 2 rows).

If we have extra space (due to a larger stride for dim 0), we can just inplace update that extra space.

Comment (shunting314) [reply torch/_inductor/lowering.py:3862]: Potentially similar optimization can be done if we want to pad with [0, 0, 0, 2] (e.g. padding 2 rows). But I don't see workloads benefit from this so far.

**Comment by shunting314** [comment]
I'm debugging the CI failure for test 'InplacePaddingTest::test_mutating_padding_output'. I can not repro locally so I logged into the CI machine which is an A10G GPU.

It turns out that we generate the same wrapper/kernel and use the same triton configs.. But on CI the correct tensor is different to the correct tensor I get on my dev machine even if I reset the seeds. I guess that due to different cuda runtime (my dev is using 12.0, CI is using 12.1).

If I initialize input data with some fixed pattern: e.g. x[i][j] = sin(i + j), y[i][j] = cos(i - j) . Then the test in CI passes. 

I guess CI just generated a set of inputs that cause larger numeric churn for matmul. I'll just enlarger the tolerance to make it pass.

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/lowering.py:3862]: Do you know how the non-dense tensor is produced ? Why are we not just computing the constant_pad_nd shape in the forward ?

Comment (shunting314) [reply torch/_inductor/lowering.py:3862]: > Do you know how the non-dense tensor is produced ? 

yea. This is generated by 'comprehensive padding' for the benefits to speed up the computation of grad_softmax_input (aka. grad_matmul_output). 

> Why are we not just computing the constant_pad_nd shape in the forward ?

Because the tensor being padded is a gradient.

In an idea world, if we are not doing mm padding in such an early stage (pattern matcher), maybe we can skip some mm padding (when we know we are doing comprehensive padding) to make the optimization looks more straightforward.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -3814,12 +3815,91 @@ def loader(idx):
...
+        return True  # be conservative
+
+    if _padding_can_be_fused():
+        return None
+
+    # Only handle 2D case for now
+    if len(padding) != 4 or len(x.get_size()) != 2:
+        return None
+
+    if not ir.is_storage_and_layout(x):
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/lowering.py:3854]: I'm not sure actually. This function is written in a way that we don't slow things down if we can not do the inplace-padding.  I'm worried if 'realize' here will slow down some cases? Or at lease increase the number of SchedulerNode even if it does not impact perf since we can fuse in scheduler anyways.

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -3814,12 +3815,91 @@ def loader(idx):
...
+        1. if the downstream op is a sum, then there is little benefit to
+           do inplace padding
+        2. if the downstream op is a matmul, doing inplace padding can
+           save membw.
+        """
+        current_node = V.graph.current_node
+        if current_node is None:
+            return True  # be conservative
+        users = tuple(current_node.users)
+        if len(users) == 1 and users[0].target in (
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/lowering.py:3839]: hmm, I think it's a bit different but correct me if I'm wrong.

What I understand is, `needs_realized_inputs` means that the inputs of the op needs to be realized.  What I need here is to check if the 'padding' can be fused to the user op. I don't have a good way to do that. Right now I just assume we can not fuse the padding into mm/addmm. But I can add convolution, bmm as well.

What do you think?

File: torch/_inductor/lowering.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/lowering.py:3856]: At this point we've already checked that it has a single, unfusible use so I dont think so.. but potentially wed want to freeze to a non-dense layout so that we could do the optimization.. in any case should be fine as is for now

[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/lowering.py:3856]: yea, good point

**Comment by etaf** [comment]
@shunting314 Can you also resolve the issue #143479 before reland?

**Review by eellison — APPROVED** [review]

File: test/inductor/test_inplace_padding.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        M, N = 2048, 2048
+        x = rand_strided((M, N), (N + 10, 1), device=GPU_TYPE)
+        y = torch.randn(N + 8, M, device=GPU_TYPE)
+        # 1e-3 tolerance may fail on CI A10G GPU.
+        check_model(self, f, (x, y), atol=1e-2, rtol=1e-2)
+
+        self.assertEqual(num_inplace_padding(), 0)
+
+    @unittest.skipIf(
+        torch.cuda.is_available()
```
Comment (etaf) [inline test/inductor/test_inplace_padding.py:121]: Can you resolve the issue #143479 here? Since there is no `CUBLAS_WORKSPACE_CONFIG` on other GPU.

Comment (shunting314) [reply test/inductor/test_inplace_padding.py:121]: I've resolve this in a local commit. Will update after resolve a few remaining tiny things for fixing the failed internal model reported.

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68340248).

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68340248).

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68340248).

File: test/inductor/test_inplace_padding.py (hunk 4 above)
Comment (etaf) [reply test/inductor/test_inplace_padding.py]: Thanks^^
